[clang][ARM64EC] Add support for hybrid_patchable attribute.

cjacek opened the pull request:
This adds support for `hybrid_patchable` on top of LLVM part from #92965 (so it depends on #92965 and this PR is meant only for the second commit). For the most part, it just adds LLVM attribute whenever C/C++ attribute is specified.

I added a warning when it's used on static function. Due to the way it works (emitting a reference to an undefined symbol that linker substitutes with a generated thunk), it can't work on static functions. MSVC just silently ignores it; using it like that seems like a non-obvious mistake to me, so I followed MSVC by allowing it too, but emitting an additional warning.

The patch does nothing special for function inlining, which matches my experimentation with MSVC. `hybrid_patchable` functions may be specified on inline functions. Depending on optimizations taken, when they are actually inlined, it has no effect, but when they are not inlined, it works as expected.

cjacek committed:
[clang][ARM64EC] Add support for hybrid_patchable attribute.

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4027,6 +4030,12 @@ def SelectAny : InheritableAttr {
 }
 
+def HybridPatchable : DeclOrTypeAttr, TargetSpecificAttr<TargetArm64EC> {
+  let Spellings = [Declspec<"hybrid_patchable">, GCC<"hybrid_patchable">];
```
I don't see any evidence that GCC supports this attribute, were you aiming for supporting `__attribute__((hybrid_patchable))` and not `[[gnu::hybrid_patchable]]`? If so, that would be a `GNU` spelling instead of a `GCC` spelling, but we typically would use a `Clang` spelling so that `[[clang::hybrid_patchable]]` works.

cjacek replied:
Yes, I was thinking mostly about `__attribute__((hybrid_patchable))` as it's the form that mingw will mostly likely want to use, so I meant `GNU`. `Clang` sounds good to me if proffered, I changed that and added a test for it.

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -477,6 +477,9 @@ def TargetELF : TargetSpec {
 def TargetELFOrMachO : TargetSpec {
   let ObjectFormats = ["ELF", "MachO"];
 }
+def TargetArm64EC : TargetSpec {
```
Should this say `Windows` somewhere in the name to make it more clear that this isn't an ARM64 attribute in general?

cjacek replied:
I renamed it to `TargetWindowsArm64EC` in the new version.

AaronBallman reviewed (commented):
At some point, you should also add documentation to clang/docs/ReleaseNotes.rst so that users know about the new functionality.

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6886,6 +6886,13 @@ static void checkAttributesAfterMerging(Sema &S, NamedDecl &ND) {
+    if (!ND.isExternallyVisible()) {
+      S.Diag(Attr->getLocation(),
+             diag::warn_attribute_hybrid_patchable_non_extern);
+      ND.dropAttr<SelectAnyAttr>();
```
```suggestion
      ND.dropAttr<HybridPatchableAttr>();
```

cjacek replied:
Good point, it was a copoy&paste typo. Actually, we don't really need it as LLVM part ignores the attribute for static functions itself. I removed that line and left only diagnostics part here to slightly simplify the code.

cjacek commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4027,6 +4030,12 @@ def SelectAny : InheritableAttr {
   let SimpleHandler = 1;
 }
 
+def HybridPatchable : DeclOrTypeAttr, TargetSpecificAttr<TargetArm64EC> {
```
Right, I don't remember why I did it like that. I changed it in the new version.

cjacek commented:
> At some point, you should also add documentation to clang/docs/ReleaseNotes.rst so that users know about the new functionality.

I included release notes changelog in the new version. I also rebased it on top of merged #92965 and addressed other comments. Thanks for reviews!

efriedma-quic reviewed (approved):
LGTM

(Please cherry-pick to 19; I'd like to have this in 19.  And maybe https://github.com/cjacek/llvm-project/commit/4febef20ab27ef4440e9dccf3cc65dd6be1f4d7c, if you wouldn't mind opening a pull request for that.)

cjacek commented:
/cherry-pick ea98dc8b8f508b8393651992830e5e51d3876728

cjacek commented:
/cherry-pick ea98dc8b8f508b8393651992830e5e51d3876728
